File: 2874_maximum_value_of_an_ordered_triplet_ii.cpp
class Solution {
public:
    long long maximumTripletValue(vector<int>& w) {
        int n = w.size();
        vector<int> rightMax(n);
        rightMax[n - 1] = w[n - 1];
        for (int i = n - 2; i >= 0; --i)
            rightMax[i] = max(rightMax[i + 1], w[i]);

        long long ans = 0;
        int leftMax = w[0];
        for (int j = 1; j < n - 1; ++j) {
            ans = max(ans, 1LL * (leftMax - w[j]) * rightMax[j + 1]);
            leftMax = max(leftMax, w[j]);
        }
        return ans;
    }
};

File: 2874_maximum_value_of_an_ordered_triplet_ii.java
class Solution {
    public long maximumTripletValue(int[] w) {
        int n = w.length;
        int[] rightMax = new int[n];
        rightMax[n - 1] = w[n - 1];
        for (int i = n - 2; i >= 0; i--)
            rightMax[i] = Math.max(rightMax[i + 1], w[i]);

        long ans = 0;
        int leftMax = w[0];
        for (int j = 1; j < n - 1; j++) {
            ans = Math.max(ans, (long) (leftMax - w[j]) * rightMax[j + 1]);
            leftMax = Math.max(leftMax, w[j]);
        }
        return ans;
    }
}
